go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -364,16 +364,41 @@ func (s *VtctldServer) ApplyVSchema(ctx context.Context, req *vtctldatapb.ApplyV
 		vs = req.VSchema
 	}
 
+	ksVs, err := vindexes.BuildKeyspace(vs, s.ws.SQLParser())
 	if req.DryRun { // we return what was passed in and parsed, rather than current
 		return &vtctldatapb.ApplyVSchemaResponse{VSchema: vs}, nil
 	}
 
-	_, err = vindexes.BuildKeyspace(vs, s.ws.SQLParser())
 	if err != nil {
 		err = vterrors.Wrapf(err, "BuildKeyspace(%s)", req.Keyspace)
 		return nil, err
 	}
 
+	// Attach unknown Vindex params to the response.
+	unknownVindexParams := make(map[string]*vtctldatapb.ApplyVSchemaResponse_ParamList)
+	var vdxNames []string
+	for name := range ksVs.Vindexes {
+		vdxNames = append(vdxNames, name)
+	}
+	sort.Strings(vdxNames)
+	for _, name := range vdxNames {
+		vdx := ksVs.Vindexes[name]
+		if val, ok := vdx.(vindexes.ParamValidating); ok {
+			ups := val.UnknownParams()
+			if len(ups) == 0 {
+				continue
+			}
+			unknownVindexParams[name] = &vtctldatapb.ApplyVSchemaResponse_ParamList{Params: ups}
+		}
+	}
+
+	if req.DryRun { // we return what was passed in and parsed, rather than current
+		return &vtctldatapb.ApplyVSchemaResponse{
+			VSchema:             vs,
+			UnknownVindexParams: unknownVindexParams,
+		}, nil
+	}
```

Comment:
We never get here do we? Looks like we return if it's a dry-run above. Maybe it's also set in BuildKeyspace which is why it was moved up?